src/lib/HST/Feature/GuardElimination.hs

```diff
@@ -1,255 +1,269 @@
-
-toDecl :: S.Pat a -> S.Exp a -> S.Decl a
-toDecl (S.PVar _ name) e = S.FunBind
+generateLet vs err gExps = do
+  varNames <- replicateM (length gExps + 1) (freshName genericFreshPrefix)
+  let startVarExpr : nextVarExprs = map S.var varNames
+      ifExprs = zipWith (rhsToIf . gExpRhs) gExps nextVarExprs
+  ifExprs' <- mapM applyGEExp ifExprs
+  let exprPats  = map (zip vs . gExpPats) gExps
+      caseExprs = zipWith3 generateNestedCases ifExprs' nextVarExprs exprPats
+      decls     = zipWith makeVarBinding varNames (caseExprs ++ [err])
+  return $ S.Let S.NoSrcSpan (S.BDecls S.NoSrcSpan decls) startVarExpr
+
+-- | Creates a function declaration for a @let@ binding of a variable with the
+--   given name to the given expression.
+makeVarBinding :: S.Name a -> S.Exp a -> S.Decl a
+makeVarBinding name expr = S.FunBind
   S.NoSrcSpan
-  [S.Match S.NoSrcSpan name [] (S.UnGuardedRhs S.NoSrcSpan e) Nothing]
-toDecl _ _ = error "GuardElimination.toDecl: Variable pattern expected"
+  [S.Match S.NoSrcSpan name [] (S.UnGuardedRhs S.NoSrcSpan expr) Nothing]
 
--- Folds the list of GExps to declarations.
-foldGEqs
-  :: Member Fresh r
-  => [S.Pat a]                   -- fresh variables for the case exps
-  -> ([S.Decl a], S.Pat a)       -- startcase ([], first generated Pattern)
-  -> [GExp a]                    -- list of pattern + rhs pair
-  -> Sem r ([S.Decl a], S.Pat a) -- a list of declarations for the let binding
-                                          -- and a free Variable for the error case
-foldGEqs vs = foldM (createDecl vs)
-
--- Generates a varbinding and a new variable for the next var binding
-createDecl
-  :: Member Fresh r
-  => [S.Pat a]                   -- generated variables
-  -> ([S.Decl a], S.Pat a)       -- (current decls , variable for let binding)
-  -> GExp a                      -- pairs of pattern to match against and a guarded Rhs
-  -> Sem r ([S.Decl a], S.Pat a) -- var bindings , variable for next match
-createDecl vs (decl, p) (ps, rhs) = do
-  nVar <- freshVarPat genericFreshPrefix
-  let varExp = CA.translatePVar nVar
-  iexp <- rhsToIf rhs varExp
-  let cexp  = createCase iexp varExp (zip vs ps)
-  let ndecl = toDecl p cexp
-  return (ndecl : decl, nVar)
-
--- TODO refactor to higher order
--- Generates a recursive case expression for each variable and pattern pair
-createCase
-  :: S.Exp a              -- ifThenElse
-  -> S.Exp a              -- the other pattern (in case pattern match or guard fails)
-  -> [(S.Pat a, S.Pat a)] -- Patterns to match (PVar , Pattern)
+-------------------------------------------------------------------------------
+-- @case@ Expressions                                                        --
+-------------------------------------------------------------------------------
+
+-- | Generates nested case expression for each variable and pattern pair.
+--
+--   @'generateNestedCases' e f [(x₁, p₁), …, (xₙ, pₙ)]@ produces an expression
+--   of the following form.
```

Comment:
```suggestion
--   of the following form:
```

Reply:
If we used a colon here, the sentence would never end with a period.

Reply:
I don't think this is a problem. I've found [this article](https://www.grammarbook.com/punctuation/colons.asp) and looking at the first example of rule 3, it seems to be correct as well.

Reply:
The article also states that "there are no hard and fast rules, except be consistent". The rule that I've adopted from @ichistmeinname (and which we should stick to for consistency's sake) is that a colon is only used if the sentence still ends with a period eventually.

Reply:
Okay, but I don't think this is a sensible rule if it prevents the usage of colons in cases like this. The only option to add a colon here would be to add a period after the example, which would look weird.
The choice between a colon and a period is a stylistic one in the end. I would definitely choose the colon in this case (and I'm [not the only one](https://english.stackexchange.com/questions/96286/period-or-colon)), but I'm fine with a period as well.